demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+                                     uint8_t * label,
+                                     mbedtls_pk_context * mbedPkContext );
+
+
+/**
+ * @brief Import the specified private key into storage.
+ *
+ * @param[in] session The PKCS #11 session.
+ * @param[in] privateKey The private key to store, in PEM format.
+ * @param[in] privateKeyLength The length of the key, including null terminator.
+ * @param[in] label The label to store the key.
+ */
+static CK_RV provisionPrivateKey( CK_SESSION_HANDLE session,
+                                  uint8_t * privateKey,
+                                  size_t privateKeyLength,
+                                  uint8_t * label );
+
+/**
+ * @brief Import the specified X.509 client certificate into storage.
+ *
+ * @param[in] session The PKCS #11 session.
+ * @param[in] certificate The certificate to store, in PEM format.
+ * @param[in] certificateLength The length of the certificate, including null terminator.
+ * @param[in] label The label to store the certificate.
+ */
+static CK_RV provisionCertificate( CK_SESSION_HANDLE session,
+                                   uint8_t * certificate,
+                                   size_t certificateLength,
+                                   uint8_t * label );
+
+/**
+ * @brief Read the specified ECDSA public key into the MbedTLS ECDSA context.
+ *
+ * @param[in] p11Session The PKCS #11 session.
+ * @param[in] pEcdsaContext The context in which to store the key.
+ * @param[in] publicKey The public key to read.
+ */
+static int extractEcPublicKey( CK_SESSION_HANDLE p11Session,
+                               mbedtls_ecdsa_context * pEcdsaContext,
+                               CK_OBJECT_HANDLE publicKey );
+
+/**
+ * @brief MbedTLS callback for signing using the provisioned private key. Used for
+ * signing the CSR.
+ *
+ * @param[in] pContext Unused.
+ * @param[in] mdAlg Unused.
+ * @param[in] pHash Data to sign.
+ * @param[in] hashLen Length of #pHash.
+ * @param[out] pSig The signature
+ * @param[out] pSigLen The length of the signature.
+ * @param[in] pRng Unused.
+ * @param[in] pRngContext Unused.
+ */
+static int32_t privateKeySigningCallback( void * pContext,
+                                          mbedtls_md_type_t mdAlg,
+                                          const unsigned char * pHash,
+                                          size_t hashLen,
+                                          unsigned char * pSig,
+                                          size_t * pSigLen,
+                                          int ( *pRng )( void *, unsigned char *, size_t ),
+                                          void * pRngContext );
+
+/**
+ * @brief MbedTLS random generation callback to generate random values with
+ * PKCS #11.
+ *
+ * @param[in] pCtx Pointer to the PKCS #11 session handle.
+ * @param[out] pRandom Buffer to write random data to.
+ * @param[in] randomLength Length of random data to write.
+ */
+static int randomCallback( void * pCtx,
```

Comment:
It would be good to commonize this between the transport library and this demo

Reply:
This can be done later.